### PR TITLE
[TPU] Proper half-precision implementation for XLA

### LINKED ISSUE #18172: XLA's 16-bit precision plugin selection is incorrect
### Bug description

(I'll point to Fabric for the motivation, but this same issue exists in the Trainer)

The connector logic to choose bf16 precision with XLA is flawed:

https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/connector.py#L435-L441

Where `XLABf16Precision` basically just sets `XLA_USE_BF16=1`: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/xlabf16.py#L25

In https://pytorch.org/xla/release/2.0/, it's explained that

> XLA_USE_BF16: If set to 1, tranforms all the PyTorch Float values into BiFloat16 when sending to the TPU device. Note that when using XLA_USE_BF16=1 tensor arithmetic will be done in reduced precision and so tensors will not be accurate if accumulated over time. 

and 

> If XLA_USE_BF16 is set, then torch.float and torch.double are both bfloat16 on TPUs.

Which corresponds to what you would expect with `precision="bf16-true"`.

However, the connector logic only selects this plugin with `precision="bf16-mixed"` even though mixed precision training is not currently supported with XLA (this is tracked in https://github.com/Lightning-AI/lightning/issues/17927).

Fixing this will be a breaking change

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

`Fabric(precision="bf16-true")` will not set `XLA_USE_BF16=1`.

(the Trainer currently does not implement `precision="bf16-true"`)

### Error messages and logs

None

### Environment

Current master

### More info

We should also support the same with `XLA_USE_F16` for `16-true` precision

cc @carmocca @justusschock @awaelchli @JackCaoG @steventk-g @Liyang90

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #18172

Successful test run
https://github.com/Lightning-AI/lightning/actions/runs/5802318415/job/15728813744?pr=18213

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @JackCaoG @steventk-g @Liyang90 @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/connector.py [hunk 1]
```diff
@@ -426,18 +425,8 @@ def _check_and_init_precision(self) -> Precision:
...
-                return XLAPrecision()
-            if self._precision_input in ("16-mixed", "bf16-mixed"):
-                if self._precision_input == "16-mixed":
-                    rank_zero_warn(
-                        "You passed `Fabric(accelerator='tpu', precision='16-mixed')` but AMP with fp16"
-                        " is not supported with TPUs. Using `precision='bf16-mixed'` instead."
-                    )
-                return XLABf16Precision()
-
+        if isinstance(self.strategy, XLAStrategy):
```
Comment (awaelchli) [inline src/lightning/fabric/connector.py:428]: Todo: include single device in this check

File: src/lightning/pytorch/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -505,23 +504,15 @@ def _check_and_init_precision(self) -> PrecisionPlugin:
...
                 return HPUPrecisionPlugin(self._precision_flag)
-        if isinstance(self.accelerator, XLAAccelerator):
-            if self._precision_flag == "32-true":
-                return XLAPrecisionPlugin()
-            if self._precision_flag in ("16-mixed", "bf16-mixed"):
-                if self._precision_flag == "16-mixed":
-                    rank_zero_warn(
-                        "You passed `Trainer(accelerator='tpu', precision='16-mixed')` but AMP with fp16"
-                        " is not supported on TPUs. Using `precision='bf16-mixed'` instead."
-                    )
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/connectors/accelerator_connector.py:516]: TODO: add this back?

**Review by carmocca — APPROVED** [review]
In a follow-up we should add support for casting the model weights to half too.

File: src/lightning/pytorch/trainer/connectors/accelerator_connector.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/pytorch/trainer/connectors/accelerator_connector.py:516]: Yes, I deleted these lines because we have an error already in the xla precision plugin. Would you like a have a customized error message like the warning above specifically for "-mixed"?

**Comment by awaelchli** [comment]
> In a follow-up we should add support for casting the model weights to half too.


My interpretation was that this happens automatically as described [here](https://pytorch.org/xla/release/2.0/index.html#xla-tensors-and-bfloat16) but for completeness, we can add the conversion anyway if it doesn't hurt.

**Review by Borda — APPROVED** [review]

File: src/lightning/fabric/plugins/precision/xla.py [hunk 3]
```diff
@@ -11,20 +11,59 @@
...
+        if precision == "16-true":
+            os.environ["XLA_USE_F16"] = "1"
+            self._desired_dtype = torch.float16
+        elif precision == "bf16-true":
+            os.environ["XLA_USE_BF16"] = "1"
+            self._desired_dtype = torch.bfloat16
+        else:
+            self._desired_dtype = torch.float32
+
+    def convert_input(self, data: Any) -> Any:
```
Comment (carmocca) [inline src/lightning/fabric/plugins/precision/xla.py:62]: Shouldn't this be removed too? I don't think the Trainer tests were failing due to a bug in the Trainer. Fabric should be impacted too with this change and the Fabric test coverage probably doesn't include such failure

Comment (awaelchli) [reply src/lightning/fabric/plugins/precision/xla.py:62]: Probably yes, i can add an integration test if there is not already to find out
